pyani_plus/public_cli.py

```diff
@@ -133,50 +135,87 @@ def check_fasta(fasta: Path) -> list[Path]:
+                    session.connection().execute(text("PRAGMA data_version;")).one()[0]
+                )
+            except OperationalError:
+                # Probably another process is updating the DB, try again soon
+                continue
+            if old_db_version == db_version:
+                # No changes, check again soon
+                continue
+            old_db_version = db_version
+            new = run.comparisons().count() - already_done - done
+            progress.update(task, advance=new)
+            done += new
+    session.close()
+
+
+def run_snakemake_with_progress_bar(  # noqa: PLR0913
     workflow_name: str,
+    database: Path,
+    run_id: int,
     targets: list[Path],
     params: dict,
     working_directory: Path,
     interval: float = 0.5,
-) -> int:
-    """Run snakemake with a progress bar.
-
-    Currently this works by monitoring the filesystem for the appearance of
-    the output files, while snakemake runs in a subprocess. This is not ideal,
-    and ideally we would use some kind of callbacks from snakemake - perhaps
-    using their logging mechanism.
-    """
+) -> None:
+    """Run snakemake with a progress bar."""
     runner = snakemake_scheduler.SnakemakeRunner(workflow_name)
     # All rest replaces one line, runner.run_workflow(targets, params, workdir=working_directory)
 
-    pending = [Path(_) for _ in targets]
-    p = Process(
-        target=runner.run_workflow,
-        args=(targets, params),
-        kwargs={"workdir": working_directory},
+    # As of Python 3.8 onwards, the default on macOS ("Darwin") is "spawn"
+    # As of Python 3.12, the default of "fork" on Linux triggers a deprecation warning.
+    # This should match the defaults on Python 3.14 onwards.
+    # Note mypy currently can't handle this dynamic situation, their issue #8603
+    p = multiprocessing.get_context(  # type:ignore [attr-defined]
+        "spawn" if sys.platform == "darwin" else "forkserver"
+    ).Process(
+        target=progress_bar_via_db_comparisons,
+        args=(
+            database,
+            run_id,
+            interval,
+        ),
     )
     p.start()
-    with Progress(*progress_columns) as progress:
-        # Same length:
-        # Indexing FASTAs
-        # Comparing pairs
-        task = progress.add_task("Comparing pairs", total=len(pending))
-        while pending:
-            p.join(interval)
-            for t in pending[:]:
-                if t.is_file():
-                    # Could print(f"Done: {t}")
-                    pending.remove(t)
-                    progress.update(task, advance=1)
-            if p.exitcode is not None:
-                # Should be finished, but was it success or failure?
-                pending = []  # to break the loop
-    p.join()  # Should be immediate as should have finished
-    if p.exitcode is None:
-        msg = "Snakemake did not stop?"
-        sys.exit(msg)
-    return p.exitcode
+
+    # Call slurm!
```

Comment:
Is it always SLURM?

Reply:
no, should be call snakemake - fixed on another PR